web/utils.py

```
@@ -938,3 +942,75 @@ def geocode_address(address):
+        )
+    # If not full but reaching capacity (80% or more)
+    elif capacity_percentage >= 80:
+        Notification.objects.create(
+            user=teacher,
+            title=f"{item_type.title()} Almost Full: {title}",
+            message=f"Your {item_type} is now at {int(capacity_percentage)}% capacity ({current_count}/{max_count} spots filled).",
+            notification_type="warning",
+        )
+
+def send_notification_email(
+    user, 
```

Ruff (0.8.2)

Missing type annotation for function argument `user`

(ANN001)